<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [2]:
import pandas as pd
import psycopg2

In [3]:
# вставьте сюда параметры подключения из юнита 1. Работа с базой данных из Python 

In [4]:
import psycopg2
import pandas as pd
DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432
connection = psycopg2.connect(
   dbname=DBNAME,
   user=USER,
   host=HOST,
   password=PASSWORD,
   port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [5]:
# текст запроса

query = f'''select count(*) from vacancies
 '''
df = pd.read_sql_query(query, connection)
df

C:\Users\Valera\AppData\Local\Temp\ipykernel_12592\786534274.py:5: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


count
0  49197

## Количество вакансий равно 49197

In [6]:
# результат запроса
query = f'''select *
        from vacancies
        limit 20  
        '''
# лучше использовать f-строки — так удобнее передавать параметры
# код запроса представляет из себя строковую переменную
df = pd.read_sql_query(query, connection)
df

C:\Users\Valera\AppData\Local\Temp\ipykernel_12592\581906716.py:8: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


id                                               name  \
0   55312386                                Компьютерный Мастер   
1   55843014                            Системный администратор   
2   54525964                      Lead Java Developer to Poland   
3   54525965                      Lead Java Developer to Poland   
4   55354053  Специалист службы поддержки с техническими зна...   
5   55906072                                          Корректор   
6   55523207                   React Front-end Middle Developer   
7   55050261                                Инженер-программист   
8   55583432                            Системный администратор   
9   55312414        Ведущий специалист группы разработки систем   
10  55471342                             Ведущий программист 1C   
11  52690985                           Senior Android Developer   
12  54525996  Senior DevOps Engineer (AWS), Investment Manag...   
13  54564021                            Системный администратор   
14  54561243                                    Product manager   
15  54243916       Начальник отдела информационной безопасности   
16  55335228        Специалист технической поддержки 1С Розница   
17  54913430  Главный инженер по сопровождению сервисов (2 л...   
18  54640182  Старший менеджер по сопровождению сделок Отдел...   
19  54447854  Сотрудник службы поддержки с техническими знан...   

                                           key_skills          schedule  \
0   Пользователь ПК\tРабота в команде\tРемонт ноут...       Полный день   
1   Средства криптографической защиты информации\t...       Полный день   
2     Spring Framework\tSQL\tHibernate ORM\tJava\tGit  Удаленная работа   
3     Spring Framework\tSQL\tHibernate ORM\tJava\tGit  Удаленная работа   
4                                                None  Удаленная работа   
5   Корректура текстов\tГрамотность\tРабота в кома...       Полный день   
6        JavaScript\tTypeScript\tCSS3\tNode.js\tHTML5     Гибкий график   
7   Java SE\tSpring Framework\tSpring Boot\tGit\tS...       Полный день   
8   Linux\tАдминистрирование сетевого оборудования...       Полный день   
9   1С программирование\t1С: Предприятие 8\t1С: До...       Полный день   
10  PostgreSQL\tSQL\tLinux\t1С программирование\tM...  Удаленная работа   
11            Android\tJava\tAndroid SDK\tKotlin\tООП       Полный день   
12         Git\tAWS\tLinux\tProject management\tSCALA  Удаленная работа   
13                                               None       Полный день   
14  Управление проектами\tProject management\tHTML...  Удаленная работа   
15  Информационная безопасность\tАудит безопасност...       Полный день   
16  1С: Розница\tПользователь ПК\tНастройка ПК\tНа...       Полный день   
17           Linux\tKubernetes\tZabbix\tDocker\tNginx       Полный день   
18  Пользователь ПК\tРабота с большим объемом инфо...       Полный день   
19                                               None  Удаленная работа   

            experience           employment  salary_from  salary_to  area_id  \
0            Нет опыта     Полная занятость      64000.0        NaN     1516   
1        От 3 до 6 лет     Полная занятость          NaN        NaN      130   
2        От 3 до 6 лет     Полная занятость          NaN        NaN      160   
3        От 3 до 6 лет     Полная занятость          NaN        NaN      159   
4            Нет опыта  Частичная занятость      15000.0        NaN     1955   
5            Нет опыта     Полная занятость          NaN        NaN     2323   
6   От 1 года до 3 лет     Полная занятость          NaN        NaN     2759   
7            Нет опыта     Полная занятость      40000.0    60000.0       49   
8   От 1 года до 3 лет     Полная занятость      50000.0        NaN        3   
9        От 3 до 6 лет     Полная занятость          NaN        NaN       47   
10       От 3 до 6 лет     Полная занятость     150000.0        NaN     2813   
11       От 3 до 6 лет     Полная занятость          NaN        Na

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [7]:
# текст запроса
query = f'''select 
        count(*)
           from employers
           
        '''
# лучше использовать f-строки — так удобнее передавать параметры
# код запроса представляет из себя строковую переменную
df = pd.read_sql_query(query, connection)
df

C:\Users\Valera\AppData\Local\Temp\ipykernel_12592\276069230.py:9: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


count
0  23501

## Количество работодателей равно 23501

In [8]:
# результат запроса

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [9]:
# текст запроса
query = f'''select 
        count(*)
           from areas
           
        '''
# лучше использовать f-строки — так удобнее передавать параметры
# код запроса представляет из себя строковую переменную
df = pd.read_sql_query(query, connection)
df

C:\Users\Valera\AppData\Local\Temp\ipykernel_12592\4092917440.py:9: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


count
0   1362

### Количество регионов равно 1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [10]:
query = f'''select 
        count(*)
           from industries
           
        '''

df = pd.read_sql_query(query, connection)
df

C:\Users\Valera\AppData\Local\Temp\ipykernel_12592\2046270931.py:7: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


count
0    294

### Количество сфер деятельности равно 294

***

### Предварительный анализ данных показывает высокую востребованность IT  специалистов во всех регионах. Предлагается большое разнообразие сфер деятельности

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [11]:
# текст запроса
query = f'''select
         a.name,
         count( v.area_id ) as cnt
      from 
         vacancies as v,
         areas as a
      where v.area_id = a.id    
      group by a.name  
      order by 2 DESC
      '''
# лучше использовать f-строки — так удобнее передавать параметры
# код запроса представляет из себя строковую переменную
df = pd.read_sql_query(query, connection)
df

C:\Users\Valera\AppData\Local\Temp\ipykernel_12592\2149177200.py:14: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


name   cnt
0             Москва  5333
1    Санкт-Петербург  2851
2              Минск  2112
3        Новосибирск  2006
4             Алматы  1892
..               ...   ...
764       Тарко-Сале     1
765    Новоаннинский     1
766            Бирск     1
767           Сасово     1
768             Анна     1

[769 rows x 2 columns]

### Наибольшее количество вакансий имеют города Москва, Санкт-Петербург, Минск, Новосибирск, Алматы

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [12]:
query = f'''select 
                count(*) as cnt     
        from vacancies
        where salary_from is not NULL or salary_to is not NULL
        '''
df = pd.read_sql_query(query, connection)
df

C:\Users\Valera\AppData\Local\Temp\ipykernel_12592\799706051.py:6: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


cnt
0  24073

### Количество вакансий где заполнено хотя бы одно из двух полей с зарплатой равно 24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [13]:
# текст запроса
query = f'''select 
                ROUND(AVG(salary_from)) AS "avg_salary_from",
                ROUND(AVG(salary_to)) AS "avg_salary_to"
        from vacancies
        where salary_from is not NULL or salary_to is not NULL
        '''
df = pd.read_sql_query(query, connection)
df

C:\Users\Valera\AppData\Local\Temp\ipykernel_12592\1630156834.py:8: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


avg_salary_from  avg_salary_to
0          71065.0       110537.0

### Средняя нижняя планка зарплаты IT специалиста находится на уровне 71 тысячи рублей. Средняя верхняя планка на уровне 110 тысяч рублей.

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [14]:
query = f'''
select  distinct schedule,
        employment,
        count(*)
from vacancies
group by 1, 2
order by 3 desc
'''

df = pd.read_sql_query(query, connection)
df

C:\Users\Valera\AppData\Local\Temp\ipykernel_12592\452873180.py:10: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


schedule           employment  count
0        Полный день     Полная занятость  35367
1   Удаленная работа     Полная занятость   7802
2      Гибкий график     Полная занятость   1593
3   Удаленная работа  Частичная занятость   1312
4     Сменный график     Полная занятость    940
5        Полный день           Стажировка    569
6     Вахтовый метод     Полная занятость    367
7        Полный день  Частичная занятость    347
8      Гибкий график  Частичная занятость    312
9        Полный день     Проектная работа    141
10  Удаленная работа     Проектная работа    133
11     Гибкий график           Стажировка    116
12    Сменный график  Частичная занятость    101
13  Удаленная работа           Стажировка     64
14     Гибкий график     Проектная работа     18
15    Сменный график           Стажировка     12
16    Вахтовый метод     Проектная работа      2
17    Сменный график     Проектная работа      1

###  Посчитано количество вакансий  в зависимости от графика работы. Очевидно предпочтение работодателей иметь работников на полный рабочий день и полную занятоять.

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [15]:
# текст запроса
query = f'''
select
        distinct experience,
        count(*) as cnt
from vacancies 
group by experience
order by cnt 
'''
df = pd.read_sql_query(query, connection)
df

C:\Users\Valera\AppData\Local\Temp\ipykernel_12592\818621942.py:10: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


experience    cnt
0         Более 6 лет   1337
1           Нет опыта   7197
2       От 3 до 6 лет  14511
3  От 1 года до 3 лет  26152

### Результаты запроса показывают наибольший спрос на специалистов с небольшим опытом работы. Также много вакансий без опыта работы.

***

In [16]:
# выводы по детальному анализу вакансий

### Выводы.
Предварительный анализ показывает высокий спрос на специалистов. Особенно в крупных городах. Также предлагается высокая зарплата даже для специалитсов с небольшим опытом работы.

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [17]:
# текст запроса
query = f'''

select
    v.employer_id,
    e.name,
    count(e.name)
from vacancies as v
join employers e on v.employer_id = e.id 

group by 1, e.name
order by 3 desc
'''
df = pd.read_sql_query(query, connection)
df

C:\Users\Valera\AppData\Local\Temp\ipykernel_12592\1383727218.py:14: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


employer_id                            name  count
0             1740                          Яндекс   1933
1             2748                      Ростелеком    491
2            78638                        Тинькофф    444
3             3529                            СБЕР    428
4            39305                   Газпром нефть    331
...            ...                             ...    ...
14901      1714361         Торговый Дом Петровский      1
14902      1714498                      А2 Юнайтед      1
14903       879898                    ТСЦ Дровосек      1
14904      1714722                  ПАВЛОДАРЭНЕРГО      1
14905        23712  ЧЕТЫРЕ ЛАПЫ, Сеть зоомагазинов      1

[14906 rows x 3 columns]

### На первом месте по количеству вакансий находится Яндекс на пятом Газпромнефть.

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [18]:
# текст запроса
query = f''' 
        SELECT areas.name,
        COUNT(DISTINCT employers.id) AS employers_count,
        count(distinct vacancies.id) vacancy_cnt
FROM areas
left JOIN employers ON employers.area = areas.id
LEFT JOIN vacancies ON vacancies.area_id = areas.id
WHERE vacancies.id is null
GROUP BY 1
order by 2 desc

 '''
df = pd.read_sql_query(query, connection)
df

C:\Users\Valera\AppData\Local\Temp\ipykernel_12592\4172112829.py:14: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


name  employers_count  vacancy_cnt
0                            Россия              410            0
1                         Казахстан              207            0
2                Московская область               75            0
3                Краснодарский край               19            0
4                          Беларусь               18            0
..                              ...              ...          ...
588                        Кувандык                0            0
589                        Куйбышев                0            0
590  Куйбышево (Ростовская область)                0            0
591                        Култаево                0            0
592                        Кульсары                0            0

[593 rows x 3 columns]

### На первом месте Россия. Странно что страну назвали регионом.

In [19]:
# результат запроса
#Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём
query = f''' 
        SELECT 
                distinct a.name,
                count(distinct v.employer_id),
                count(distinct v.id)
        from 
                vacancies as v
        LEFT JOIN areas as a ON a.id=v.area_id
        GROUP BY 1
        order by 2 desc

 '''
df = pd.read_sql_query(query, connection)
df

C:\Users\Valera\AppData\Local\Temp\ipykernel_12592\2584971963.py:15: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


name  count  count
0                             Москва   2320   5333
1                    Санкт-Петербург   1451   2851
2                              Минск    851   2112
3                        Новосибирск    802   2006
4                             Алматы    798   1892
..                               ...    ...    ...
764                      Электроугли      1      1
765                           Эртиль      1      1
766                             Юрга      1      1
767  Яблоновский (Республика Адыгея)      1      2
768                               Яя      1      1

[769 rows x 3 columns]

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [20]:

query = f'''

select
    e.name,
    count(distinct v.area_id)
from employers as e
join vacancies v on v.employer_id = e.id 

group by e.name
order by 2 desc
limit 5
'''
df = pd.read_sql_query(query, connection)
df

C:\Users\Valera\AppData\Local\Temp\ipykernel_12592\3737076438.py:13: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


name  count
0                  Яндекс    181
1              Ростелеком    152
2              Спецремонт    116
3  Поляков Денис Иванович     88
4               ООО ЕФИН      71

### Яндекс имеет 181 регион где публикует вакансиию.

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [21]:
query = f'''
SELECT count(*)
FROM employers e
LEFT JOIN employers_industries ei 
ON e.id = ei.employer_id
where  industry_id is null

'''
df = pd.read_sql_query(query, connection)
df

C:\Users\Valera\AppData\Local\Temp\ipykernel_12592\2912276779.py:9: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


count
0   8419

### Сфера деятельности не указана у 8491 работодателя.

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [22]:
# текст 
query = f'''
SELECT
e.name,
 
count(*) 
FROM employers e
LEFT JOIN employers_industries ei 
ON e.id = ei.employer_id

group by e.name
having count(*) = 4
order by 1
limit 4


'''
df = pd.read_sql_query(query, connection)
df

C:\Users\Valera\AppData\Local\Temp\ipykernel_12592\3294061226.py:18: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


name  count
0  101 Интернет      4
1      21vek.by      4
2          2ГИС      4
3            2К      4

### На третьем месте в списке компаний с четырьмя видами деятельности находится 2ГИС

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [23]:
# текст запроса
query = f'''
SELECT count(*)
FROM employers e
LEFT JOIN employers_industries ei 
ON e.id = ei.employer_id
LEFT JOIN industries i 
ON i.id = ei.industry_id
where  i.name = 'Разработка программного обеспечения'

'''
df = pd.read_sql_query(query, connection)
df


C:\Users\Valera\AppData\Local\Temp\ipykernel_12592\4140091834.py:12: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


count
0   3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [25]:
# результат запроса
import urllib.parse as req


path_common = req.urlparse('https://ru.wikipedia.org/wiki/Города-миллионеры_России#Список_городов-миллионеров')
url = path_common.scheme + '://' + req.quote(path_common.netloc) +  req.quote(path_common.path) + '#'+ req.quote(path_common.fragment)
cities = pd.read_html(url)[1]

city = tuple(cities['Город'].tolist())
print(city)
query = f'''
select
        a.name as region,
        count(*) as vacancies_count
    from
        vacancies as v
        join employers as e on v.employer_id = e.id
        join areas as a on v.area_id = a.id
    group by
        v.area_id, a.name, e.id, e.name
    having
        a.name in {city} and e.name = 'Яндекс'
    

    union all
    

    select
        'Total',
        sum(x.cnt)
    from
        (
        select
            count(*) as cnt
        from
            vacancies as v
            join employers as e on v.employer_id = e.id
            join areas as a on v.area_id = a.id
        group by
            v.area_id, a.name, e.id, e.name
        having
            a.name in {city} and e.name = 'Яндекс'
        ) as x
        

    order by 2
    

'''
df = pd.read_sql_query(query, connection)
df

('Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Нижний Новгород', 'Красноярск', 'Челябинск', 'Самара', 'Уфа', 'Ростов-на-Дону', 'Краснодар', 'Омск', 'Воронеж', 'Пермь', 'Волгоград')


C:\Users\Valera\AppData\Local\Temp\ipykernel_12592\1607225273.py:50: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


region  vacancies_count
0              Омск             21.0
1         Челябинск             22.0
2        Красноярск             23.0
3         Волгоград             24.0
4    Ростов-на-Дону             25.0
5            Казань             25.0
6             Пермь             25.0
7               Уфа             26.0
8            Самара             26.0
9         Краснодар             30.0
10          Воронеж             32.0
11      Новосибирск             35.0
12  Нижний Новгород             36.0
13     Екатеринбург             39.0
14  Санкт-Петербург             42.0
15           Москва             54.0
16            Total            485.0

***

In [ ]:
# выводы по анализу работодателей
query = f'''

SELECT  COUNT(*) 
FROM areas a
JOIN vacancies v
ON v.area_id = a.id
JOIN employers e
ON v.employer_id = e.id
WHERE e.name = 'Яндекс' and a.name in {city}
GROUP BY a.name

'''
df = pd.read_sql_query(query, connection)
df

C:\Users\Valera\AppData\Local\Temp\ipykernel_12880\552892087.py:14: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


count
0      24
1      32
2      39
3      25
4      30
5      23
6      54
7      36
8      35
9      21
10     25
11     25
12     26
13     42
14     26
15     22

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [ ]:
# текст запроса

In [ ]:
# результат запроса

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [ ]:
# текст запроса

In [ ]:
# результат запроса

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [ ]:
# текст запроса

In [ ]:
# результат запроса

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [ ]:
# текст запроса

In [ ]:
# результат запроса

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [ ]:
# текст запроса

In [ ]:
# результат запроса

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [ ]:
# текст запроса

In [ ]:
# результат запроса

***

In [ ]:
# выводы по предметному анализу

# Общий вывод по проекту

In [ ]:
# подведем итог исследования, обобщите выводы
# здесь можно (это будет плюсом) провести дополнительные исследования данных, сделать прогнозы, продумать варианты продолжения исследования